### ISSUE TITLE
Does it make sense to treat elements in a tuple/array literal as part of the @parameter object?

### ISSUE THREAD

**Issue by IndianBoy42** [nvim-treesitter/nvim-treesitter-textobjects#86]
They are syntactically similar, and semantically not that different. I think there should be a text object for entries in lists/tuples/dictionaries, but the question is whether it should be in the parameter text object or another (@entry? @element)

**Comment by stsewd** [comment]
@IndianBoy42 do you have a file we can use to test what you mean?

**Comment by snezhniylis** [comment]
This can be a problem for dynamically typed programming languages such as Python and Lua where dicts/arrays/... can be passed directly as arguments.
```python
function_call(1, [8, 38, 2, 5])
```

```lua
function_call({
  value1 = true,
  some_number = 35,
  child_table = {
    something = "something"
  }
}, true)
```

I find it very convincing to be able to select such objects. This way, I can quickly swap arguments, move some to their own local variables, etc.

But I also frequently find it lacking that there's no selector for dict entries / table members out of the box.

**Comment by stsewd** [comment]
This was already done in https://github.com/nvim-treesitter/nvim-treesitter-textobjects/pull/87/. But I agree that it makes sense to have those as a separate capture.

**Comment by IndianBoy42** [comment]
I think arguments could be made either way.

Either you just want a 'smarter' select inside comma separated list, or you really want to be specific about the semantic element

One if the troubles is the sheer number of different text objects you could define semantically

**Comment by melkster** [comment]
I would also love to have a text object for comma separated items inside an array/list/table/dictionary. To me it makes sense to have a different capture group than `@parameter`, perhaps it could be called `@field` or `@listItem`?

**Comment by yochem** [comment]
I would go with `@item` or `@element` I think.

**Comment by adueck** [comment]
What's the state on this? Is there a text object for items/elements in an array/list? I would love to be able to select and swap items in an array.

```ts
// before
const arr = ["foo", { a: true }];
```
```ts
// after
const arr = [{ a: true }, "foo"];
```

**Comment by kiyoon** [comment]
@adueck Nothing has been formally decided but many languages like lua and python adopted this with `@parameter`. I think with JS it could be implemented to be consistent if you wish.

**Comment by adueck** [comment]
Hi @kiyoon, thanks very much for the response and explanation. Yes, I would love to have this (array elements) implemented under `@parameter` just like with lua and python.

In fact, being able to swap array elements semantically with tree-sitter was the whole reason that I made the switch from VSCode -> NeoVim. 😅 If this was possible using `@parameter` it would be wonderful.

**Comment by kiyoon** [comment]
@adueck Sure, i will take a look when I have time.

**Comment by adueck** [comment]
@kiyoon, thanks so much. 🙇‍♂️ Really love [your](https://github.com/kiyoon/haskell-scope-highlighting.nvim) [other work](https://github.com/kiyoon/treemux) btw

**Comment by kiyoon** [comment]
@adueck When you mean by spread operator, do you mean something like this?

```javascript
const newArray = [0, ...array, 4];
```

I think whatever is inside the element it should just select the whole element, so without further work it should work with the spread operators.

However, there might be a couple of things missing in the initial implementation, thus we mey need to tune it a couple of times.

**Comment by kiyoon** [comment]
> Really love your [other](https://github.com/kiyoon/haskell-scope-highlighting.nvim) [work](https://github.com/kiyoon/treemux) btw 

Thanks, I'm glad! Feel free to file any issues because I don't use haskell myself so it's good to know more feedback.

**Comment by mawkler** [comment]
@kiyoon 

> Nothing has been formally decided

I'm curious why nothing has been decided yet. To me it makes sense to be able to differenciate between `@parameter` and something like `@element` because sometimes one can contain the other causing ambiguity. And because they are two different concepts.

Is it because a name hasn't been decided on yet? Or what is the next step here? 🙂

**Comment by yochem** [comment]
Just FYI, this:

```
$XDG_CONFIG_HOME/nvim/queries/ecma/textobjects.scm
```

Should probably be:

```
$XDG_CONFIG_HOME/nvim/after/queries/ecma/textobjects.scm
```
In order to extend the original highlights. And of course the ft of your JS file then has to be 'ecma'.

**Comment by adueck** [comment]
> [@adueck](https://github.com/adueck) When you mean by spread operator, do you mean something like this?
> 
> ```js
> const newArray = [0, ...array, 4];
>```

Yes, exactly @kiyoon. So if I go in there and swap the 1st and 2nd elements it would change to:

```js
const newArray = [...array, 0, 4];
```

As you said, just keeping the spread as part of the array element.

@mawkler, I agree it could be nice to have something like `@element`, but for now I think it would be great to bring JS/TS up to speed with how these parameters/elements can be manipulated in Lua and Python. The precision of accessing elements vs parameters might be nice, but I don't think it's too much of a trouble to go to the parameter, then go inside the parameter array and then work with the "parameters" of the array. In this way it's like a nested structure, like working with lists of lists.

```js
//      params  1      2                
//              v      v
function myFunc("foo", ["bar", "baz")) {
//                      ^      ^
//               params 1      2  
//   ...
}
```

To me, it does make sense to have array and set elemens under the same semantic umbrella as `@parameter`, being an element that fills out a structure. Just like a `@parameter` fills out a function statement, an argument fills out a function call, an element fills out an array, and element fills out a set, an entry fills out an object. This is consistent with how it is implemented in Python, using it for dictionaries, tuples, lists, function definitions, function calls.

<img width="816" alt="Image" src="https://github.com/user-attachments/assets/5fe435de-a68d-423c-ae2b-84ade5fb430a" />

I'm not saying it wouldn't be nice to have a specific `@element` textobject, but I see how it makes sense to use `@parameter` for both, and I personally like using one object/keybinding for both.

**Comment by kiyoon** [comment]
> I'm curious why nothing has been decided yet. To me it makes sense to be able to differenciate between `@parameter` and something like `@element` because sometimes one can contain the other causing ambiguity. And because they are two different concepts.
> 
> Is it because a name hasn't been decided on yet? Or what is the next step here? 🙂

I guess they are two different concepts but in some languages they are used as a similar concept like putting a table as an argument. Also, they are different thus they don't collide with each other. Making too many separate names can cause overwhelmingly many keybindings causing unintuitive experience.. We follow what the community wants and so far it happens to be combining them into `@parameter`.

**Comment by kiyoon** [comment]
The feature has been added to the ECMA script with lists and objects. Let me know if there are more cases that needs to be addressed.

**Comment by adueck** [comment]
Awesome, thanks @kiyoon, it's working great for me so far. So nice to see that this functionality I dreamed of is here!

**Comment by adueck** [comment]
> Let me know if there are more cases that needs to be addressed.

Hi @kiyoon, the only case that I've encountered that needs to be adressed is when using [destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring).

For example:

```js
// 1 and 2 are @parameter s
// but fst and snd are not
const [fst, snd] = [1, 2];

// a and b are @parameter s
// but foo and bar are not
const { foo, bar } = { a, b };
```

**Comment by kiyoon** [comment]
@adueck Done!

**Comment by adaszko** [comment]
I wanted the same thing but for Rust for a loong time!  Even built https://github.com/adaszko/tree_climber_rust.nvim as a slight generalization of that use case (after also failing with `after/queries/rust/testobjects.scm` approach too ;)  Any chance of updating the Rust grammar too? :)

**Comment by adaszko** [comment]
> @adaszko Sure. Can you provide me with all types of example snippets you could think of? I know it has macros like `vec![] ` or something like that.

With pleasure!  Will post a list later

**Comment by adaszko** [comment]
The test suite for `tree_climber_rust.nvim` can serve as a list.  There's:
 * [`tuple_expression`](https://github.com/adaszko/tree_climber_rust.nvim/blob/002358ab6f0b4696b75905804ea7f1dca34a7ccd/lua/tree_climber_rust.lua#L461)
 * [`tuple_type`](https://github.com/adaszko/tree_climber_rust.nvim/blob/002358ab6f0b4696b75905804ea7f1dca34a7ccd/lua/tree_climber_rust.lua#L595C4-L595C14)
 * [`call_expression`](https://github.com/adaszko/tree_climber_rust.nvim/blob/002358ab6f0b4696b75905804ea7f1dca34a7ccd/lua/tree_climber_rust.lua#L737C4-L737C19)
 * [`type_arguments`](https://github.com/adaszko/tree_climber_rust.nvim/blob/002358ab6f0b4696b75905804ea7f1dca34a7ccd/lua/tree_climber_rust.lua#L935C4-L935C18)
 * [`array_expression`](https://github.com/adaszko/tree_climber_rust.nvim/blob/002358ab6f0b4696b75905804ea7f1dca34a7ccd/lua/tree_climber_rust.lua#L1070C4-L1070C20)
 * [`function_item`](https://github.com/adaszko/tree_climber_rust.nvim/blob/002358ab6f0b4696b75905804ea7f1dca34a7ccd/lua/tree_climber_rust.lua#L1272C4-L1272C17)

**Comment by kiyoon** [comment]
I tested those and the only missing implementation is `array_expression`?

```rust
// tuple_expression
let t = (1, 2, 3);

// tuple_type (appears in the return type)
fn returns_tuple() -> (usize, u32, u8) {
    (42, 7, 1)
}

// call_expression
fn f(x: i32, y: i32, z: i32) -> i32 {
    x + y + z
}
let sum = f(1, 2, 3);

// type_arguments (generic type parameters)
fn g<T1, T2, T3>(_: T1, _: T2, _: T3) {}
g::<u8, u16, u32>(1, 2, 3);

// array_expression
let arr = [1, 2, 3];

// function_item (formal parameter list in the declaration)
fn h(a: u8, b: u16, c: u32) {
}
```

**Comment by mawkler** [comment]
@adaszko With all due respect I think the conversation has derailed a little from the original topic. Perhaps you could create a new issue for the Rust implementation?

@kiyoon

> Making too many separate names can cause overwhelmingly many keybindings causing unintuitive experience

I personally disagree. If the user doesn't want to have many different key-bindings they could just use the `w` motion for all text-editing. I think granularity is what make Vim/Neovim as well as nvim-treesitter-textobjects so powerful. Also, since the user has to create their own key-bindings anyway there's no stopping them from mapping both `@parameter` and `@element` to the same key.

> We follow what the community wants and so far it happens to be combining them into `@parameter`

Could you please elaborate on what you're basing this on? In what way has the community expressed that they don't want an `@element` that's separate from `@parameter`? Doesn't the existance of this whole issue thread contradict that statement?

**Comment by kiyoon** [comment]
> Also, since the user has to create their own key-bindings anyway there's no stopping them from mapping both @parameter and @element to the same key.

With this logic, a user can create their custom queries to do whatever they want anyway. See #753 .

> Could you please elaborate on what you're basing this on?

It has already been implemented in many languages. And there's no one that I remember who complained about it when implementing this feature. And what's most important is to be consistent. We can either remove the feature in Rust/Python/Lua etc. to disappoint some users or we can keep the consistency and try to implement in other languages too.

In addition, as I explained earlier, `@element` and `@parameter` have many overlaps, in many languages. In lua or JS, a table or an object and be a parameter and it's a very common convention. If we were to separate the two, the `@parameter` should select the entire table, while the `@element` should select an element within it. However, due to the convention, it wouldn't be intuitive at all because the user wanted to select the element and they are using `@parameter`.

All languages are different, and the goal of this plugin is to try to generalise this as much as possible so with one config you expect to work 90% correctly in many languages. If we were to separate all specific cases, there will be hundreds, or thousands of queries that is only used in one or two languages.

**Comment by adueck** [comment]
@kiyoon I found another case for `@parameter` that would be nice to include in TypeScript. It would be nice to have the properties in object types and interfaces treated as `@parameter`s.

```ts
type Person = {
    name: string,
    age: number,
}

// or

type Person = {
    name: string;
    age: number;
}

// or

interface Person {
    name: string;
    age: number;
}

// or

function makePerson(name: string, age: number): { name: string, age: number } {
    return { name, age };
}
```

In types it would be nice to have `name: string` to be a `@parameter` just like in arrays with keys/objects.

**Comment by kiyoon** [comment]
@adueck Thanks. Implemented the case!

**Comment by adueck** [comment]
Thanks @kiyoon you are the best. Tirelessly working to massively improve the quality of life of fellow developers!! 👏

**Comment by adueck** [comment]
@kiyoon one more idea for `@parameter` ...

Would it make sense to have each type in a union as a `@parameter` ?

```ts
type Person = Baby | Child | Adult | Senior;
```

Here the `Baby`, `Child`, `Adult` and `Senior` could be thought of like the parameters in an array. And it would be nice to delete them as such. Like if you wanted to edit that union and delete the `Child` type (delete `@parameter`) then you would get:

```ts
type Person = Baby | Adult | Senior;
```

The same could go for intersections:

```ts
type Person = HealthInfo & AddressInfo & History;
```

**Comment by kiyoon** [comment]
You're welcome!

Nice suggestion, but I've found three problems.

**1. Consistency**: No other languages have such feature so if we implement it, we should keep everything consistent.

**2. Implementation:**: The union and intersection operators are structured nestedly, unlike other types of parameters.

```scheme
(union_type
      (union_type
        (union_type
          (type_identifier)
          (type_identifier))
        (type_identifier))
      (type_identifier))
```

whilst, other parameters would be in siblings form like:

```scheme
(function_parameters
  (parameter)
  (parameter)
  (parameter))
```

Such structure make it difficult to match with pure queries. It needs some while loops in lua to match properly, but we won't be able to make such custom function just for this case.

Note that the structure can be more complicated with parentheses:

```ts
type Person = Baby | (Child & Adult) | Senior;
```

**3. Ambiguity:** What if it was mixed union and intersection?

```ts
type Person = Baby | Child & Adult | Senior;
//                     ^ cursor, we delete @parameter.outer
```

Should we delete the union or intersection operator?

**4. Type used other than definition:** If it's in a function parameters, it shouldn't really be treated as parameter because it will ruin the original experience.

```ts
// the `Person` should not be selected
function greet(person: Person): string {
  return `Hello, ${person.name}`;
}
```

**Comment by kiyoon** [comment]
That said, I've made a reference implementation:

```scheme
;; Support for union and intersection types as @parameter
;; type Person = Baby | (Child & Adult) | Senior;
((type_identifier) @parameter.inner
  (#has-ancestor? @parameter.inner union_type intersection_type)
  (#has-ancestor? @parameter.inner type_alias_declaration))

; capture the outer if it is not part of a union or intersection type
(type_alias_declaration
  value: (type_identifier) @parameter.outer)

((union_type
   (type_identifier) @_start
   .
   "|" @_end
   .
   (_))
 (#has-ancestor? @_start type_alias_declaration)
 (#make-range! "parameter.outer" @_start @_end))

((union_type
   (_)
   .
   "|" @_start
   .
   (type_identifier) @_end)
 (#has-ancestor? @_end type_alias_declaration)
 (#make-range! "parameter.outer" @_start @_end))

((intersection_type
   (type_identifier) @_start
   .
   "&" @_end
   .
   (_))
 (#has-ancestor? @_start type_alias_declaration)
 (#make-range! "parameter.outer" @_start @_end))

((intersection_type
   (_)
   .
   "&" @_start
   .
   (type_identifier) @_end)
 (#has-ancestor? @_end type_alias_declaration)
 (#make-range! "parameter.outer" @_start @_end))

; if parenthesis is used alone, capture with the () as well
((parenthesized_type
   "(" @_start
   .
   (type_identifier)
   .
   ")" @_end)
 (#has-ancestor? @_end type_alias_declaration)
 (#make-range! "parameter.outer" @_start @_end))

; Incomplete scenario: type Person = | Child;
; then it will just have one type identifier.
(
(union_type
  .
   (type_identifier) .) @parameter.outer
 (#has-ancestor? @parameter.outer type_alias_declaration)
  )

(
(intersection_type
  .
   (type_identifier) .) @parameter.outer
 (#has-ancestor? @parameter.outer type_alias_declaration)
  )
```

Feel free to put in your `~/.config/nvim/after/queries/typescript/textobjects.scm` with `;; extends` statement on top and test it out.

The implementation may not be very accurate, expect bugs.

**Comment by adueck** [comment]
Wow @kiyoon thanks very much for the good explanation and reference implementation.

**Comment by kiyoon** [comment]
One more issue:

**5: Easy to break:** unlike other statements like `(a, b, c,)  // not a syntax error`, it's easy to break the grammar while deleting the type unions.

```ts
type Person = Baby |;
type Person = | Baby;
```

In such case, the parsing will not work properly and the last element might not be deleted. Note that in the reference implementation, I addressed the second case but not the first case (and it may be different across languages). So while you try to delete or manipulate the statement, it will just stop working at times, hurting the user experience.

So in summary, I think it's a niche case that not everyone would agree, also adding quite a big maintenance cost. I would prefer to keep it in the personal config rather than officially supporting it.
